lxd/lifecycle/instance.go

```diff
@@ -0,0 +1,67 @@
+package lifecycle
+
+import (
+	"fmt"
+	"net/url"
+
+	"github.com/lxc/lxd/lxd/operations"
+	"github.com/lxc/lxd/lxd/project"
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/api"
+)
+
+// Internal copy of the instance interface
+type instance interface {
+	Project() string
+	Operation() *operations.Operation
+	IsSnapshot() bool
+}
+
+// InstanceAction represents a lifecycle event action for instances
+type InstanceAction string
+
+// All supported lifecycle events for instances
+const (
+	InstanceCreated   = InstanceAction("created")
+	InstanceStarted   = InstanceAction("started")
+	InstanceStopped   = InstanceAction("stopped")
+	InstanceShutdown  = InstanceAction("shutdown")
+	InstanceRestarted = InstanceAction("restarted")
+	InstancePaused    = InstanceAction("paused")
+	InstanceResumed   = InstanceAction("resumed")
+	InstanceRestored  = InstanceAction("restored")
+	InstanceDeleted   = InstanceAction("deleted")
+	InstanceRenamed   = InstanceAction("renamed")
+	InstanceUpdated   = InstanceAction("updated")
+)
+
+// Event creates the lifecycle event for an action on an instance
+func (a InstanceAction) Event(kind, name string, inst instance, ctx map[string]interface{}) api.EventLifecycle {
+	if inst.IsSnapshot() {
+		kind = "snapshot"
+	}
+	eventType := fmt.Sprintf("instance-%s", a)
+	u := fmt.Sprintf("/1.0/instances/%s", url.PathEscape(name))
+
+	if kind != "instance" {
```

Comment:
As kind is used for logic we should have a constant and type defined for it so we can easily find out the possible values.

Reply:
Ah, I removed the constants after your previous comment about not needing another InstanceType so as to not make it as concrete. But you're right, it is indeed concrete so I'll add it back.